### PR TITLE
V3events

### LINKED ISSUE #4259: Should Event API follow CNCF's Cloud Events
We should reconsider standardizing our Event API to follow CNCF's https://cloudevents.io/ standard. 

Benefits: better alignment with the space, standardized format which aligns better with best practices 
Cons: IDEs will need to restructure their clients

Thoughts?

### PR DESCRIPTION
**Pull request by baleshk**
Fixes: #4259  https://github.com/GoogleContainerTools/skaffold/issues/4259

**Description**
In this pull request, we have added support for cloudevents standards on a different route /v3/events. This is the first draft of the changes, gives a good idea about the changes required for implementing this feature. Code is in buiding state and functionality can be tested. Although message schema is not finalised and not all fields are populated with correct values for now. In addition to code changes, inputs are required on finalising the message schema, mechanism for delivering these events (which is another route, /v3/events).
Some hacks are added to make sure build is passing and will be removed after this first review cycle. All test cases are not passing and new ones are not yet added. Complete unit test coverage will be provided in next review request.

### CODE REVIEW

**Comment by aaron-prindle** [comment]
For generating the protos, would it be possible to make it so that [hack/generate-protos.sh](https://github.com/GoogleContainerTools/skaffold/blob/main/hack/generate-proto.sh) works with the changes here?  It seems in this branch some files from v1 and v2 that the script uses were removed and the v3 changes were not integrated with the script there.

File: pkg/skaffold/output/output.go [hunk 1]
```diff
@@ -25,15 +25,17 @@ import (
...
 const timestampFormat = "2006-01-02 15:04:05"
 
 type skaffoldWriter struct {
-	MainWriter  io.Writer
-	EventWriter io.Writer
-	task        constants.Phase
-	subtask     string
+	MainWriter   io.Writer
+	EventWriter  io.Writer
+	EventWriter1 io.Writer
```
Comment (aaron-prindle) [inline pkg/skaffold/output/output.go]: nit: EventWriterV3 might make more sense as a name

File: pkg/skaffold/server/server.go [hunk 2]
```diff
@@ -190,7 +203,19 @@ func newGRPCServer(preferredPort int, usedPorts *util.PortSet) (func() error, in
 }
 
 func newHTTPServer(preferredPort, proxyPort int, usedPorts *util.PortSet) (func() error, error) {
-	mux := runtime.NewServeMux(runtime.WithProtoErrorHandler(errorHandler), runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{OrigName: true, EmitDefaults: true}))
+	mux := runtime.NewServeMux(
```
Comment (aaron-prindle) [inline pkg/skaffold/server/server.go:204]: Do we still need the `runtime.WithProtoErrorHandler(errorHandler)` to be part of our ServeMux?  It seems this snippet is from - https://grpc-ecosystem.github.io/grpc-gateway/docs/development/grpc-gateway_v2_migration_guide/#we-now-use-the-camelcase-json-names-by-default, just want to make sure we dont' lose any options/behavior that was present in the past marshalling from this config

File: pkg/skaffold/server/server.go [hunk 3]
```diff
@@ -230,7 +260,7 @@ type errResponse struct {
 }
 
 func errorHandler(ctx context.Context, _ *runtime.ServeMux, marshaler runtime.Marshaler, writer http.ResponseWriter, _ *http.Request, err error) {
-	writer.Header().Set("Content-type", marshaler.ContentType())
+	//writer.Header().Set("Content-type", marshaler.ContentType())
```
Comment (aaron-prindle) [inline pkg/skaffold/server/server.go]: nit: remove commented code

File: pkg/skaffold/server/v3/server.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+type Server struct {
+	BuildIntentCallback  func()
+	SyncIntentCallback   func()
+	DeployIntentCallback func()
+	AutoBuildCallback    func(bool)
+	AutoSyncCallback     func(bool)
+	AutoDeployCallback   func(bool)
+}
+
+// TODO(marlongamez): Add Set*Callback() funcs once going for v1 feature parity
```
Comment (aaron-prindle) [inline pkg/skaffold/server/v3/server.go]: nit: remove comment

File: pkg/skaffold/output/output.go (hunk 1 above)
Comment (baleshk) [reply pkg/skaffold/output/output.go]: Fixed.

File: pkg/skaffold/server/server.go (hunk 2 above)
Comment (baleshk) [reply pkg/skaffold/server/server.go:204]: Fixed. ErrorHandler removed. It is no more required in the version 2. We are still using the similar marshalling options to OrigName, EmitDefaults, ie. UseProtoNames and EmitUnpopulated

File: pkg/skaffold/server/server.go (hunk 3 above)
Comment (baleshk) [reply pkg/skaffold/server/server.go]: Fixed.

File: pkg/skaffold/server/v3/server.go (hunk 4 above)
Comment (baleshk) [reply pkg/skaffold/server/v3/server.go]: Fixed.

**Comment by baleshk** [comment]
> For generating the protos, would it be possible to make it so that [hack/generate-protos.sh](https://github.com/GoogleContainerTools/skaffold/blob/main/hack/generate-proto.sh) works with the changes here? It seems in this branch some files from v1 and v2 that the script uses were removed and the v3 changes were not integrated with the script there.

Balesh: Fixed

**Pull request merged**